### PR TITLE
upgrade from 0.5.0 to 3.1.0

### PR DESCRIPTION
**Pull request by ankit1910**
rewrite javascript using prototype approach
make changes in model
update stylesheets

### CODE REVIEW

File: app/models/spree/product_decorator.rb [hunk 1]
```diff
@@ -1,31 +1,23 @@
 Spree::Product.class_eval do
 
-  def option_values
-    @_option_values ||= Spree::OptionValue.for_product(self).order(:position).sort_by {|ov| ov.option_type.position }
-  end
+  has_many :option_values, -> { uniq }, through: :variants
```
Comment (nishant-cyro) [inline app/models/spree/product_decorator.rb:3]: @ankit1910 - We don't require to add uniq here.

File: app/models/spree/product_decorator.rb [hunk 2]
```diff
@@ -1,31 +1,23 @@
...
 
   def grouped_option_values
-    @_grouped_option_values ||= option_values.group_by(&:option_type)
-  end
-
-  def variants_for_option_value(value)
-    @_variant_option_values ||= variants.includes(:option_values).all
-    @_variant_option_values.select { |i| i.option_value_ids.include?(value.id) }
+    option_types.includes(:option_values).order(:position)
```
Comment (nishant-cyro) [inline app/models/spree/product_decorator.rb]: @ankit1910 - Why we have eager loaded option_values in grouped_option_values?

File: app/models/spree/product_decorator.rb [hunk 3]
```diff
@@ -1,31 +1,23 @@
...
-    return @_variant_options_hash if @_variant_options_hash
-    hash = {}
-    variants.includes(:option_values).each do |variant|
-      variant.option_values.each do |ov|
-        otid = ov.option_type_id.to_s
-        ovid = ov.id.to_s
-        hash[otid] ||= {}
-        hash[otid][ovid] ||= {}
-        hash[otid][ovid][variant.id.to_s] = variant.to_hash
+  def variants_option_value_details
```
Comment (nishant-cyro) [inline app/models/spree/product_decorator.rb:9]: @ankit1910 - Lets refactor this.

File: vendor/assets/javascripts/frontend/variant_options.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
+var OptionValuesHandler = function(selectors) {
```
Comment (nishant-cyro) [inline vendor/assets/javascripts/frontend/variant_options.js]: @ankit1910 - Add this class in a namescape.

File: vendor/assets/javascripts/frontend/variant_options.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      _this.resetAllNextLevel($this);
+      _this.unlockNextLevel($this);
+
+      if($this.data('level') == options["option_type_count"]) {
+        _this.setVariantWithSelecetedValues();
+      }
+    }
+  });
+
+  this.clearButtons.on('click', function(e) {
```
Comment (nishant-cyro) [inline vendor/assets/javascripts/frontend/variant_options.js:48]: @ankit1910 - Lets move the logic in methods.

File: vendor/assets/javascripts/frontend/variant_options.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    _this.updateSiblings($(this));
+    _this.resetAllNextLevel($(this));
+  })
+};
+
+OptionValuesHandler.prototype.disableCartInputFields = function(value) {
+  this.addToCartButton.prop('disabled', value);
+  this.quantityField.prop('disabled', value);
+
+  if(value) { this.priceHeading.html('Select Variant'); }
```
Comment (nishant-cyro) [inline vendor/assets/javascripts/frontend/variant_options.js:60]: @ankit1910 - I guess we should move this value in a data attribute.

File: app/models/spree/product_decorator.rb (hunk 2 above)
Comment (nishant-cyro) [inline app/models/spree/product_decorator.rb]: @ankit1910 - Please rename this method.

File: app/models/spree/product_decorator.rb (hunk 1 above)
Comment (ankit1910) [inline app/models/spree/product_decorator.rb:3]: it is required

**Pull request merged**